File: pipeline/renderer_final.py
import os
from pathlib import Path
import tempfile
import time
import hashlib
from urllib.parse import unquote
import json
import boto3

import ffmpeg

# first step to ensure we have all parts
# then call process()


def main(args, context):

    # Get the service client.
    s3_client = boto3.client('s3')

    key = unquote(args['Records'][0]['s3']['object']['key'])

    # parse the key
    choir_id, song_id, def_id, run_id, _, rows_hash = parse_key(key)

    src_bucket = os.environ['SRC_BUCKET']

    # Check all parts present if, not abort
    key_prefix = f'{choir_id}+{song_id}+{def_id}+{run_id}'
    contents = s3_client.list_objects(
        Bucket=src_bucket,
        Prefix=key_prefix
    )
    row_keys = [x['Key'] for x in contents.get('Contents', [])
                if x['Size'] > 0]

    # Sort to make sure we are in correct order
    row_keys.sort(key=lambda x: int(parse_key(x)[4]))

    # Calc hash of found parts to make sure we have all, if not abort
    if calc_hash_of_keys(row_keys) != rows_hash:
        # add choir_id/song_id/status for render status updates
        # if we arrive here, not all parts are rendered yet, so status =
        # 'rendered'
        ret = {
            'status': 'missing rows',
            'choir_id': choir_id,
            'song_id': song_id}
        return ret

    process_args = {}
    process_args['key'] = key
    process_args['row_keys'] = row_keys
    r = process(process_args)
    # render status data
    # if we arrive here, all parts are rendered, so status = 'composited'

    #lambda_client = boto3.client('lambda')

    #ret = {"choir_id": choir_id,
    #       "song_id": song_id,
    #       "status": "composited"}

    #lambda_client.invoke(
    #    FunctionName=os.environ['STATUS_LAMBDA'],
    #    Payload=json.dumps(ret),
    #    InvocationType='Event'
    #)

def process(args):
    # Get the service client.
    s3_client = boto3.client('s3')

    key = args['key']

    # parse the key
    choir_id, song_id, def_id, run_id, row_num, _ = parse_key(key)

    src_bucket = os.environ['SRC_BUCKET']
    dst_bucket = os.environ['DEST_BUCKET']
    # misc_bucket = os.environ['MISC_BUCKET']

    # Download the definition file for this job
    #definition_bucket = os.environ['DEF_BUCKET']
    #definition_key = f'{choir_id}+{song_id}+{def_id}.json'
    #definition_object = s3_client.get_object(
    #    Bucket=definition_bucket,
    #    Key=definition_key,
    #)
    #definition = json.load(definition_object['Body'])
    #output_spec = definition['output']

    row_keys = args['row_keys']

    ###
    # Combine video and audio
    ###

    # video
    if len(row_keys) > 1:
        # Multiple video parts
        video_parts = []
        audio_parts = []
        for row_key in row_keys:
            _, _, _, _, row_num, _ = parse_key(row_key)
            row_url = s3_client.generate_presigned_url(
                ClientMethod='get_object',
                Params={
                    'Bucket': src_bucket,
                    'Key': row_key
                }
            )
            row_part = ffmpeg.input(row_url,
                                    seekable=0,
                                    thread_queue_size=64)
            if row_num != -1:
                video_parts.append(row_part.video)
            audio_parts.append(row_part.audio)

        video = ffmpeg.filter(video_parts, 'vstack',
                              inputs=len(video_parts))
        audio = ffmpeg.filter(audio_parts,
                              'amix',
                              dropout_transition=180,
                              inputs=len(audio_parts))
    else:
        # Just a single video part
        row_key = row_keys[0]
        _, _, _, _, row_num, _ = parse_key(row_key)
        row_url = s3_client.generate_presigned_url(
            ClientMethod='get_object',
            Params={
                'Bucket': src_bucket,
                'Key': row_key
            }
        )
        row_part = ffmpeg.input(row_url,
                                seekable=0,
                                thread_queue_size=64)
        if row_num != -1:
            video = row_part.video
        else:
            video = None
        audio = row_part.audio

    # Output
    output_key = f'{choir_id}+{song_id}+{def_id}-preprod.nut'
    # output_url = get_output_url(output_key)

    kwargs = {}
    if 'duration' in args:
        kwargs['t'] = int(args['duration'])

    if 'loglevel' in args:
        kwargs['v'] = args['loglevel']

    tempfile.tempdir = '/mnt/tmp'
    with tempfile.TemporaryDirectory() as tmp:
        # join temp directory with our filename
        path = os.path.join(tmp, output_key)

        pipeline = ffmpeg.output(audio,
                                 video,
                                 path,
                                 format='nut',
                                 pix_fmt='yuv420p',
                                 acodec='pcm_s16le',
                                 vcodec='mpeg2video',
                                 r=25,
                                 seekable=0,
                                 qscale=1,
                                 qmin=1,
                                 **kwargs
                                 )

        cmd = pipeline.compile()
        print("ffmpeg command to run: ", cmd)
        t1 = time.time()
        pipeline.run()
        t2 = time.time()

        # upload temp file to S3
        s3_client.upload_file(path, dst_bucket, output_key)

    ret = {'dst_key': output_key,
           'run_id': run_id,
           'def_id': def_id,
           'render_time': int(t2 - t1),
           'status': 'merged'}

    return ret


def parse_key(key):
    choir_id, song_id, def_id, run_id, section_id = Path(key).stem.split('+')
    row_num, rows_hash = section_id.split('@')
    return choir_id, song_id, def_id, run_id, int(row_num), rows_hash


def calc_hash_of_keys(keys):
    rows = [int(parse_key(x)[4]) for x in keys]
    return calc_hash_rows(rows)


def calc_hash_rows(rows):
    val = '-'.join([str(x) for x in sorted(rows)])
    h = hashlib.sha1(val.encode('utf-8')).hexdigest()
    return h[:8]

File: pipeline/renderer_final.js
// environment variables
const LOCAL_MODE = !!process.env.LOCAL_MODE
const DEST_BUCKET = process.env.DEST_BUCKET
const LOCAL_BUCKETS_PATH = '../buckets'

// node modules
const fs = require('fs')
const crypto = require('crypto')
const path = require('path')
const aws = require('aws-sdk')
const tmpdir = require('tmpdir')
const ffmpeg = require('fluent-ffmpeg')
const ffmpegrunner = require('ffmpegrunner')
const run = ffmpegrunner.run
const glob = require('glob')

// aws objects
const S3 = new aws.S3()

// sha1 hash
const sha1 = (str) => {
  const shasum = crypto.createHash('sha1')
  shasum.update(str)
  return shasum.digest('hex')
}

// break file name into constituent bits
const parseKey = (key) => {
  // split the incoming key into bits
  const parsed = path.parse(key).name.split('+')
  const pk = {
    choirId: parsed[0],
    songId: parsed[1],
    defId: parsed[2],
    runId: parsed[3]
  }
  const bits = parsed[4].split('@')
  pk.rowNum = bits[0]
  pk.rowsHash = bits[1]
  return pk
}

// build the filter graph for processing audio & video
const buildComplexFilter = (videos) => {
  // complex filter
  const filters = []
  const vparts = []
  const aparts = []
  let f
  for (const i in videos) {
    const video = videos[i]
    video.id = i.toString()

    // audio only videos are on row -1
    if (video.rowNum !== '-1') {
      vparts.push(video.id + ':v')
    }
    aparts.push(video.id + ':a')
  }
  // video pipeline
  f = {
    inputs: vparts,
    filter: 'vstack',
    options: {
      inputs: vparts.length
    },
    outputs: 'v'
  }
  filters.push(f)
  f = {
    inputs: aparts,
    filter: 'amix',
    options: {
      inputs: aparts.length,
      dropout_transition: 180

    },
    outputs: 'a'
  }
  filters.push(f)
  // add our filter graph to the command
  return {
    filters: filters,
    outputs: ['v', 'a']
  }
}

// main
const main = async (event, context) => {
  console.log('renderer_final')

  let key, bucket

  // when running locally
  if (LOCAL_MODE) {
    // expect 'event' to be a flat object
    key = event.key
    bucket = event.bucket
  } else {
    // event is an AWS event
    key = unescape(event.Records[0].s3.object.key)
    bucket = event.Records[0].s3.bucket.name
  }

  // split the incoming key into bits
  const pk = parseKey(key)

  // check to see if all the finalparts are present
  const keyPrefix = `${pk.choirId}+${pk.songId}+${pk.defId}+${pk.runId}`
  let rowKeys
  if (LOCAL_MODE) {
    const p = path.join(LOCAL_BUCKETS_PATH, bucket, keyPrefix + '*')
    rowKeys = glob.sync(p)
  } else {
    const response = S3.listObjects({ Bucket: bucket, Prefix: keyPrefix }).promise()
    rowKeys = response.Contents.map((obj) => { return obj.Key })
  }
  rowKeys.sort()

  // get list of run ids
  const videos = rowKeys.map(rk => parseKey(rk))
  const rows = videos.map((v) => { return v.rowNum })
  console.log(rowKeys, rows)
  const hash = sha1(rows.join('-')).substr(0, 8)
  console.log('incoming hash', pk.rowsHash, 'calculated hash', hash)
  if (hash !== pk.rowsHash) {
    return { ok: false, msg: 'not all rows here yet' }
  }

  // build ffmpeg command
  const command = ffmpeg()

  // add the inputs
  for (const i in rowKeys) {
    let geturl
    const rk = rowKeys[i]
    if (LOCAL_MODE) {
      geturl = rk
    } else {
      geturl = await S3.getSignedUrlPromise('getObject', { Bucket: bucket, Key: rk })
    }
    command.addInput(geturl)
      .inputOptions(['-seekable 0', '-thread_queue_size 64'])
  }

  // build the video & audio pipelines
  const complexFilter = buildComplexFilter(videos)
  command.complexFilter(complexFilter.filters, complexFilter.outputs)

  // create temporary directory
  const tmppath = tmpdir.createTmpDirectory()
  const outputKey = `${pk.choirId}+${pk.songId}+${pk.defId}-preprod.nut`
  const outpath = path.join(tmppath, outputKey)

  // set output parameters
  command
    .output(outpath)
    .outputFormat('nut') // nut container
    .outputOptions([
      '-pix_fmt yuv420p',
      '-acodec pcm_s16le', // PCM audio
      '-vcodec mpeg2video', // mpeg2video video
      //        '-preset fast', // fast
      '-r 25', // 25 fps
      //        '-qscale 1', // ?
      '-qmin 1'])
  await run(command, true)

  // copy the temporary file to output bucket
  if (LOCAL_MODE) {
    const desturl = path.join(LOCAL_BUCKETS_PATH, DEST_BUCKET, outputKey)
    fs.copyFileSync(outpath, desturl)
  } else {
    // upload to S3
    await S3.putObject({
      Bucket: DEST_BUCKET,
      Key: outputKey,
      Body: fs.createReadStream(outpath)
    }).promise()
  }

  tmpdir.removeTmpDirectory(tmppath)

  return { ok: true }
}

exports.main = main
